src/main/kotlin/com/teambition/kafka/connect/mongo/database/DatabaseReader.kt

```diff
@@ -112,12 +112,12 @@ class DatabaseReader(val uri: String,
         mongoCollection
             .find()
             .batchSize(batchSize)
-            .filter(Filters.gt("_id", offsetId))
-            .sort(Document("_id", 1))
+            .filter(Filters.gt("createdAt", offsetId))
+            .sort(Document("createdAt", 1))
```

Comment:
How is he data in the collection modified? Usually `createdAt` is set on initial create but not updated on modification. Is there an `updatedAt` column? I'm not sure how the collection behaves. This may be good to go. Just asking to confirm.

Reply:
If that's indeed the right field then this looks good to me. In a _future_ and separate change we could make it configurable to work with any collection.

Reply:
Yes there is an `updatedAt` column in our catalog schema, but I'm modelling this change on a 1-for-1 swap from `_id` to `createdAt`, both of which have a "write-once" semantic that does not ever change the value and generally values will always increase in terms of sort position.

The mechanism as it exists is pretty basic.

- Phase 1: initial import
  - Sort the collection, publish to kafka in batches, tracking offest for pause/resume
- Phase 2: tail the oplog

I think for both `_id` and `createdAt` there's failure cases where we'll skip documents that looks like this:

- Do initial import to completion
  - Final offset gets stored in kafka
- Stop the KC source
- Insert a document with an older `_id` or `createdAt` (lower than the saved offset)
  - If KC source is running during this insert, the oplog tailing will process this document OK
- Start the KC source
- In this case, that document will not match our initial import offset and it will remain unsynced unless we start a fresh sync manually
- To solve this, we'd need a more robust already-imported-tracking technique, but exactly what that might be could be tricky (bloom filter?)
- The same failure can happen in this case as well
  - import is running at offset X but oplog tailing hasn't started yet
  - document is inserted with createdAt less than X

Reply:
I amended this to rename `objectId` to `sortField` to be a bit more generic as well, but not yet configurable.

Reply:
Here's the [documentation for the JDBC source connector's incremental query modes](https://docs.confluent.io/current/connect/kafka-connect-jdbc/source-connector/index.html#incremental-query-modes).

> In **[timestamp column mode]**, a single column containing a modification timestamp is used to track the last time data was processed and to query only for rows that have been modified since that time. 

Wouldn't the equivalent of timestamp-column configuration work if we were to use the `updatedAt` field as the last-modified timestamp? If the connector is stopped then we need to find records that have been changed since the last committed timestamp.


It's true that a "late-arriving" record would be missed. But that'll only happen on a direct, out-of-band db modification because code sets the `updatedAt` field?

The **timestamp and incrementing columns** is designed to to be more complete. But, as you suggest, that won't be an option for us because we don't have incrementing id's.

Reply:
They are totally serial and I don't think the code as written is able to handle updates that get through the oplog during import. So far I'm not convinced this really handles anything other than the happy and quiet path without some omissions.

Reply:
Put another way, I think I'll probably need to study what JDBC does in timestamp column mode and just mirror that with the same caveats.

Reply:
Easy enough to switch to `updatedAt` and I confirmed our Catalog documents have both `createdAt` and `updatedAt` set to the some value on the initial insert, so that simplifies things.

Reply:
> What happens if Phase 1 takes a while and records are modified during that time?

I'm 95% certain the code as is will not sync those modifications. It's a single mongodb query and I think the cursor basically computes the matching documents once and will not append to that set as the database continues to change. 

I could address that by putting the query in a while loop until there are no results, indicating the sync has "caught up" and then switching to oplog tailing (starting from the updatedAt timestamp of the most recent document we synced)

> What happens if the connector is stopped and the last known oplog offset is out of range? (ie, the oplog subscription isn't valid.)

The "offset" stored is not an actual oplog offset or kafka offset, it's just the cimestamp value of `updatedAt` so there's no real "out of range", just increasingly large data sets to sync as time proceeds away from that timestamp.

Reply:
OK I did another read through the code and documented it [here in the README](https://github.com/reactioncommerce/kafka-connect-mongo/blob/37d3f8872b33e13bbf3852807c3e3270f3b2014a/README.md#how-the-mongo-source-connector-works-implementation-notes).

> What happens if Phase 1 takes a while and records are modified during that time?

**Updated answer**: Since the oplog tailing and collection import both start from the same offset timestamp, writes during import will be synced during oplog tailing. Only caveat would be in fairly unlikely bad circumstances (huge data set to import, slowness getting it into kafka), if the mongodb oplog has rolled past that timestamp we'd lose those writes. For now I think this mechanism is plenty robust for our needs.

> What happens if the connector is stopped and the last known oplog offset is out of range? (ie, the oplog subscription isn't valid.)

**Updated answer**: The initial import will start from that timestamp and sync the current state of those documents, then we'll process however much of the oplog is available after that timestamp until we're caught up.

The only glitch I can see with this would be for inserts and updates that happen during the downtime, as I think those documents would "flicker" like this:

- current state would sync to kafka from the collection
- oplog tailing would find the insert and sync the initial version
  - Search data would be stale during this window
- oplog tailing would find the update and again sync the current version